File: scripts/build-full-archive.mjs
import fs from 'node:fs';
import path from 'node:path';

const SOURCE_BASE='https://raw.githubusercontent.com/arsazet17/pozitron-top3-v1.0/main/';
const SEARCH_PARTS=[
  ['js','top3-search-pack-01.js'],['js','top3-search-pack-02.js'],['js','top3-search-pack-03.js'],['js','top3-search-pack-04.js'],
  ['js','top3-search-mini-001.js'],['js','top3-search-mini-002.js'],['js','top3-search-mini-003.js'],['js','top3-search-mini-004.js'],
  ['js','top3-search-mini-005.js'],['js','top3-search-mini-006.js'],['js','top3-search-mini-007.js'],['js','top3-search-mini-008.js'],
  ['js','top3-search-mini-009.js'],['js','top3-search-mini-010.js'],['js','top3-search-mini-011.js'],['js','top3-search-mini-012.js'],
  ['js','top3-search-tail-001.js'],['js','top3-search-tail-002.js'],['js','top3-search-tail-003.js'],['js','top3-search-tail-004.js'],
  ['js','top3-search-rest-001.js'],
  ['js','top3-search-bulk-001.js'],['js','top3-search-bulk-002.js'],
  ['raw','top3-search-bulk-003a.txt'],['raw','top3-search-bulk-003b.txt'],['raw','top3-search-bulk-003c.txt'],
  ['js','top3-search-bulk-004.js'],
  ['raw','top3-search-bulk-005a.txt'],['raw','top3-search-bulk-005b.txt'],['raw','top3-search-bulk-005c.txt'],
  ['js','top3-search-bulk-006.js'],
  ['raw','top3-search-bulk-007p1.txt'],['raw','top3-search-bulk-007p2a.txt'],['raw','top3-search-bulk-007p2b.txt'],['raw','top3-search-bulk-007p2c.txt'],['raw','top3-search-bulk-007p2d.txt'],['raw','top3-search-bulk-007p2e.txt'],['raw','top3-search-bulk-007p3.txt'],['raw','top3-search-bulk-007p4.txt'],['raw','top3-search-bulk-007p5.txt'],['raw','top3-search-bulk-007p6.txt'],
  ['js','top3-search-bulk-008.js'],['js','top3-search-bulk-009.js'],['js','top3-search-bulk-010.js'],['js','top3-search-bulk-011.js'],['js','top3-search-bulk-012.js'],['js','top3-search-bulk-013.js'],
  ['js','top3-search-pack-final.js']
];
const EARLY_FROM=11, EARLY_TO=247986, EARLY_COUNT=247976, DATED_FROM=247987;
const OUT_DIR='data/full-archive';

function readJSON(file,fallback){try{return JSON.parse(fs.readFileSync(file,'utf8'))}catch{return structuredClone(fallback)}}
function pad2(n){return String(n).padStart(2,'0')}
function addMinutesStamp(date,time,minutes){
  const [y,m,d]=String(date).split('-').map(Number),[hh,mm]=String(time).split(':').map(Number);
  const x=new Date(Date.UTC(y,m-1,d,hh,mm)+minutes*60000);
  return {date:`${x.getUTCFullYear()}-${pad2(x.getUTCMonth()+1)}-${pad2(x.getUTCDate())}`,time:`${pad2(x.getUTCHours())}:${pad2(x.getUTCMinutes())}`};
}
function decodePacked(meta){const s=String(meta?.data||''),out=[];for(let i=0;i+2<s.length;i+=3)out.push(s.slice(i,i+3));return out}
function expandBootstrap(meta){
  if(!meta?.data)return[];
  const combos=decodePacked(meta),out=[];
  for(let i=0;i<combos.length;i++){
    const stamp=i===0?meta.first:addMinutesStamp(meta.regularStart.date,meta.regularStart.time,(i-1)*Number(meta.stepMinutes||30));
    const combo=combos[i];
    out.push({draw:String(Number(meta.fromDraw)+i),date:stamp.date,time:stamp.time,A:+combo[0],B:+combo[1],C:+combo[2],combo});
  }
  if(meta.count!=null&&out.length!==Number(meta.count))throw new Error(`bootstrap count mismatch ${out.length} != ${meta.count}`);
  return out;
}
async function fetchText(name){
  const r=await fetch(SOURCE_BASE+name,{cache:'no-store'});
  if(!r.ok)throw new Error(`${name}: HTTP ${r.status}`);
  return r.text();
}
async function loadEarlyArchive(){
  const sandbox={TOP3_SEARCH_PACK:''};
  for(const [kind,name] of SEARCH_PARTS){
    const text=await fetchText(name);
    if(kind==='raw'){
      sandbox.TOP3_SEARCH_PACK+=text.trim();
    }else{
      new Function('window','atob','Uint8Array','Date','Object',text)(sandbox,atob,Uint8Array,Date,Object);
    }
  }
  const a=sandbox.TOP3_SEARCH_ARCHIVE;
  if(!a)throw new Error('TOP3_SEARCH_ARCHIVE was not created');
  if(a.count!==EARLY_COUNT||a.newestId!==EARLY_TO||a.oldestId!==EARLY_FROM)throw new Error(`early archive boundary mismatch: ${a.oldestId}..${a.newestId}, count=${a.count}`);
  if(a.codeById(EARLY_TO)!=='655'||a.codeById(EARLY_FROM)!=='699')throw new Error(`early archive control mismatch: newest=${a.codeById(EARLY_TO)} oldest=${a.codeById(EARLY_FROM)}`);
  const out=new Array(EARLY_COUNT);
  for(let id=EARLY_FROM;id<=EARLY_TO;id++){
    const c=a.codeById(id);
    if(!/^\d{3}$/.test(String(c||'')))throw new Error(`bad early combo №${id}: ${c}`);
    out[id-EARLY_FROM]=c;
  }
  return out;
}
function loadDatedTail(){
  const archive=readJSON('data/archive.json',[]),bootstrap=expandBootstrap(readJSON('data/bootstrap-tail.json',null));
  const byDraw=new Map();
  for(const r of [...archive,...bootstrap]){
    const id=Number(r?.draw),combo=String(r?.combo||'');
    if(Number.isInteger(id)&&/^\d{3}$/.test(combo))byDraw.set(id,{...r,draw:String(id),combo});
  }
  const ids=[...byDraw.keys()].filter(x=>x>=DATED_FROM).sort((a,b)=>a-b);
  if(!ids.length||ids[0]!==DATED_FROM)throw new Error(`dated archive must start at №${DATED_FROM}; got ${ids[0]}`);
  const last=ids.at(-1);
  const combos=[];
  for(let id=DATED_FROM;id<=last;id++){
    const r=byDraw.get(id);
    if(!r)throw new Error(`dated archive gap at №${id}`);
    combos.push(r.combo);
  }
  return {combos,last,rows:byDraw};
}

const early=await loadEarlyArchive();
const dated=loadDatedTail();
const combos=[...early,...dated.combos];
const expected=dated.last-EARLY_FROM+1;
if(combos.length!==expected)throw new Error(`full archive count mismatch ${combos.length} != ${expected}`);

fs.mkdirSync(OUT_DIR,{recursive:true});
const meta={
  schema:3,
  source:'TOP-3 full archive · verified early search pack + dated archive',
  earlySourceRepo:'arsazet17/pozitron-top3-v1.0',
  order:'oldest-to-newest',
  fromDraw:EARLY_FROM,
  toDraw:dated.last,
  total:combos.length,
  datedFromDraw:DATED_FROM,
  datedToDraw:dated.last,
  datedCount:dated.combos.length,
  undatedCount:early.length,
  data:combos.join('')
};
fs.writeFileSync(path.join(OUT_DIR,'all.json'),JSON.stringify(meta));
fs.writeFileSync(path.join(OUT_DIR,'unique.json'),JSON.stringify([...new Set(combos)].sort()));
console.log(`FULL ARCHIVE OK: №${meta.fromDraw}…№${meta.toDraw}; total=${meta.total}; early=${early.length}; dated=${dated.combos.length}; first=${combos[0]}; last=${combos.at(-1)}`);

File: scripts/build_full_archive.py
#!/usr/bin/env python3
"""Build compact browser archive from the repository XLSX without third-party packages.

The historical spreadsheet may contain either a single 3-digit combination column
or separate A/B/C digit columns. Old rows without date/time are kept as real draw
number + combination only; no dates are invented.
"""
import json
import os
import re
import zipfile
from pathlib import Path
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta

ROOT = Path(__file__).resolve().parents[1] if "scripts" in Path(__file__).parts else Path.cwd()
DEFAULT_XLSX = ROOT / "data" / "TOP-3_архив_A_B_C_ИСПРАВЛЕННЫЙ_компактно.xlsx"
XLSX = Path(os.getenv("TOP3_FULL_ARCHIVE_XLSX", str(DEFAULT_XLSX)))
BOOTSTRAP = ROOT / "data" / "bootstrap-tail.json"
DATED_ARCHIVE = ROOT / "data" / "archive.json"
OUT_DIR = ROOT / "data" / "full-archive"
FROM_DRAW = 11

NS = "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}"


def read_json(path, fallback):
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return fallback


def shared_strings(zf):
    try:
        with zf.open("xl/sharedStrings.xml") as fh:
            root = ET.parse(fh).getroot()
    except KeyError:
        return []
    out = []
    for si in root.findall(f"{NS}si"):
        out.append("".join(t.text or "" for t in si.iter(f"{NS}t")))
    return out


def cell_value(cell, shared):
    typ = cell.get("t")
    if typ == "inlineStr":
        return "".join(t.text or "" for t in cell.iter(f"{NS}t"))
    v = cell.find(f"{NS}v")
    if v is None or v.text is None:
        return ""
    raw = v.text.strip()
    if typ == "s":
        try:
            return shared[int(raw)]
        except Exception:
            return raw
    return raw


def as_digit(text):
    text = str(text or "").strip()
    return str(int(float(text))) if re.fullmatch(r"\d(?:\.0+)?", text) else None


def combo_from_cells(cells):
    b = str(cells.get("B", "")).strip()
    if re.fullmatch(r"\d{3}", b):
        return b
    for cols in (("A","B","C"),("B","C","D")):
        ds = [as_digit(cells.get(c, "")) for c in cols]
        if all(d is not None for d in ds):
            return "".join(ds)
    for col, text in cells.items():
        text = str(text).strip()
        if col != "A" and re.fullmatch(r"\d{3}", text):
            return text
    a = str(cells.get("A", "")).strip()
    if len(cells) == 1 and re.fullmatch(r"\d{3}", a):
        return a
    return None


def read_combos(path):
    if not path.exists():
        raise FileNotFoundError(path)
    combos = []
    with zipfile.ZipFile(path) as zf:
        shared = shared_strings(zf)
        sheet_name = "xl/worksheets/sheet1.xml"
        if sheet_name not in zf.namelist():
            candidates = sorted(n for n in zf.namelist() if n.startswith("xl/worksheets/sheet") and n.endswith(".xml"))
            if not candidates:
                raise RuntimeError("В XLSX не найден лист с архивом")
            sheet_name = candidates[0]
        with zf.open(sheet_name) as fh:
            for event, elem in ET.iterparse(fh, events=("end",)):
                if elem.tag != f"{NS}row":
                    continue
                cells = {}
                for c in elem.findall(f"{NS}c"):
                    ref = c.get("r", "")
                    m = re.match(r"([A-Z]+)", ref)
                    if m:
                        cells[m.group(1)] = cell_value(c, shared)
                combo = combo_from_cells(cells)
                if combo is not None:
                    combos.append(combo)
                elem.clear()
    if len(combos) < 1000:
        raise RuntimeError(f"Слишком мало комбинаций в XLSX: {len(combos)}")
    return combos


def expand_bootstrap(meta):
    if not isinstance(meta, dict) or not meta.get("data"):
        return []
    data = str(meta["data"])
    combos = [data[i:i+3] for i in range(0, len(data), 3) if len(data[i:i+3]) == 3]
    out = []
    start_draw = int(meta["fromDraw"])
    first = meta["first"]
    regular = datetime.strptime(f"{meta['regularStart']['date']} {meta['regularStart']['time']}", "%Y-%m-%d %H:%M")
    step = int(meta.get("stepMinutes", 30))
    for i, combo in enumerate(combos):
        if i == 0:
            ds, tm = first["date"], first["time"]
        else:
            stamp = regular + timedelta(minutes=(i-1)*step)
            ds, tm = stamp.strftime("%Y-%m-%d"), stamp.strftime("%H:%M")
        out.append({"draw": str(start_draw+i), "date": ds, "time": tm, "combo": combo})
    if meta.get("count") is not None and len(out) != int(meta["count"]):
        raise RuntimeError(f"bootstrap count mismatch: {len(out)} != {meta['count']}")
    return out


def orientation_score(combos, known_rows):
    score = 0
    checked = 0
    to_draw = FROM_DRAW + len(combos) - 1
    for row in known_rows:
        try:
            draw = int(row.get("draw"))
        except Exception:
            continue
        combo = str(row.get("combo", ""))
        if FROM_DRAW <= draw <= to_draw and re.fullmatch(r"\d{3}", combo):
            checked += 1
            if combos[draw - FROM_DRAW] == combo:
                score += 1
    return score, checked


def main():
    raw_combos = read_combos(XLSX)
    bootstrap = sorted(expand_bootstrap(read_json(BOOTSTRAP, {})), key=lambda x: int(x.get("draw", 0)))
    dated = read_json(DATED_ARCHIVE, [])
    known_rows = [*dated, *bootstrap]
    forward = raw_combos
    reverse = list(reversed(raw_combos))
    fs, fc = orientation_score(forward, known_rows)
    rs, rc = orientation_score(reverse, known_rows)
    if rs > fs:
        combos = reverse
        orientation = "sheet-newest-to-oldest -> reversed"
        score, checked = rs, rc
    elif fs > rs:
        combos = forward
        orientation = "sheet-oldest-to-newest"
        score, checked = fs, fc
    else:
        combos = reverse
        orientation = "fallback newest-to-oldest -> reversed"
        score, checked = rs, rc
    if checked and score == 0:
        raise RuntimeError(f"Не удалось согласовать порядок XLSX с известными тиражами: checked={checked}")
    print(f"XLSX orientation: {orientation}; matched={score}/{checked}")
    base_to = FROM_DRAW + len(combos) - 1

    for row in bootstrap:
        draw = int(row["draw"])
        combo = str(row["combo"])
        if not re.fullmatch(r"\d{3}", combo):
            raise RuntimeError(f"Некорректная bootstrap-комбинация №{draw}: {combo}")
        if draw <= base_to:
            pos = draw - FROM_DRAW
            if pos < 0 or combos[pos] != combo:
                raise RuntimeError(
                    f"Контроль архива не прошёл на №{draw}: XLSX={combos[pos] if 0 <= pos < len(combos) else 'нет'}, bootstrap={combo}"
                )
            continue
        if draw != FROM_DRAW + len(combos):
            raise RuntimeError(f"Разрыв полного архива: после №{FROM_DRAW + len(combos) - 1} пришёл №{draw}")
        combos.append(combo)
        base_to = draw

    dated_by_draw = {}
    for row in [*dated, *bootstrap]:
        try:
            draw = int(row.get("draw"))
        except Exception:
            continue
        if row.get("date") and row.get("time") and re.fullmatch(r"\d{3}", str(row.get("combo", ""))):
            dated_by_draw[draw] = row

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    meta = {
        "schema": 2,
        "source": "TOP-3 full archive · repository XLSX + verified dated tail",
        "order": "oldest-to-newest",
        "fromDraw": FROM_DRAW,
        "toDraw": FROM_DRAW + len(combos) - 1,
        "total": len(combos),
        "datedFromDraw": min(dated_by_draw) if dated_by_draw else None,
        "datedToDraw": max(dated_by_draw) if dated_by_draw else None,
        "datedCount": len(dated_by_draw),
        "undatedCount": len(combos) - len(dated_by_draw),
        "data": "".join(combos),
    }
    (OUT_DIR / "all.json").write_text(json.dumps(meta, ensure_ascii=False, separators=(",", ":")), encoding="utf-8")
    (OUT_DIR / "unique.json").write_text(json.dumps(sorted(set(combos)), ensure_ascii=False, separators=(",", ":")), encoding="utf-8")
    print(
        f"FULL ARCHIVE OK: {len(combos)} combos, №{meta['fromDraw']}…№{meta['toDraw']}; "
        f"dated={meta['datedCount']}; unique={len(set(combos))}"
    )


if __name__ == "__main__":
    main()
